flax/optim/base.py

```diff
@@ -463,7 +465,20 @@ def __init__(self, *traversals_and_optimizers):
 
   def init_state(self, params):
     sub_states = []
-    for traversal, opt in zip(self.traversals, self.sub_optimizers):
+    seen = {}
+    for idx, (traversal,
+              opt) in enumerate(zip(self.traversals, self.sub_optimizers)):
+
+      keykeys = traverse_util.unflatten_dict({
+        key_tuple: '/'.join(('',) + key_tuple)
+        for key_tuple in traverse_util.flatten_dict(params)
+      })
+      for key in traversal.iterate(keykeys):
+        if key in seen:
+          raise ValueError(f'Key "{key}" processed by multiple optimizers: '
+                           f'{seen[key]}, {idx}.')
+        seen[key] = idx
```

Comment:
Can't we just unflatten the params and do a raw string comparison?

```suggestion
    seen = set()
    for traversal, opt in zip(self.traversals, self.sub_optimizers):
      for key_tuple in traverse_util.flatten_dict(params):
        if key_tuple in seen:
          path = '/'.join(key_tuple)
          raise ValueError(f'Key "{path}" processed by multiple optimizers')
        seen.add(key_tuple)
```

Reply:
We only want to bail if the same key is matched multiple times by different `traversal.iterate()` - so line 476 is crucial with this test. Note that `traverse_util.Traversal` only provides an iterator over the *values* hence the creation of `keykeys`.

Reply:
Ah ok that makes sense, so since `traversal.iterate` can be different dependning on which traversal we use, we want to call it here explicitly.

Reply:
Ahh ok, that makes sense. So in that case it seems the flatted/unflatten logic is there to be able to determine whether two `ModelParamTraversal` instances yield values for the same paths in the params, is that right? (I think the API of `traverse_util.Traversal` doesn't require using a dict-like structure, this seems to be specific to `ModelParamTraversal`.)

So since that logic seems quite special for `ModelParamTraversal`, and we could get other traversal classes in the future, I wonder if it would make sense to instead add something like `traverse_util.Traversal.iterate_with_path`, which also yields the path. Then you could simply do something like this here:

```python
for path, _ in traversal.iterate_with_path(params):
  if path in seen:
    raise ValueError(f'Key "{path}" processed by multiple optimizers: '
                     f'{seen[path]}, {idx}.')
  seen[path] = idx
```

I think then you also wouldn't have to do the extra flattening/unflattening you are doing here. 

WDYT?